winch/codegen/src/abi/mod.rs

```diff
@@ -118,178 +125,310 @@ pub(crate) trait ABI {
     }
 
-    /// Allocate a new stack abi arg.
-    pub fn stack_offset(offset: u32, ty: WasmType) -> Self {
-        Self::Stack { ty, offset }
+    /// Allocate a new stack [`ABIOperand`].
+    pub fn stack_offset(offset: u32, ty: WasmType, size: u32) -> Self {
+        Self::Stack { ty, offset, size }
     }
 
-    /// Is this abi arg in a register.
+    /// Is this [`ABIOperand`] in a register.
     pub fn is_reg(&self) -> bool {
         match *self {
-            ABIArg::Reg { .. } => true,
+            ABIOperand::Reg { .. } => true,
             _ => false,
         }
     }
 
-    /// Get the register associated to this arg.
+    /// Unwraps the underlying register if it is one.
+    ///
+    /// # Panics
+    /// This function panics if the [`ABIOperand`] is not a register.
+    pub fn unwrap_reg(&self) -> Reg {
+        match self {
+            ABIOperand::Reg { reg, .. } => *reg,
+            _ => unreachable!(),
+        }
+    }
+
+    /// Get the register associated to this [`ABIOperand`].
     pub fn get_reg(&self) -> Option<Reg> {
         match *self {
-            ABIArg::Reg { reg, .. } => Some(reg),
+            ABIOperand::Reg { reg, .. } => Some(reg),
             _ => None,
         }
     }
 
-    /// Get the type associated to this arg.
+    /// Get the type associated to this [`ABIOperand`].
     pub fn ty(&self) -> WasmType {
         match *self {
-            ABIArg::Reg { ty, .. } | ABIArg::Stack { ty, .. } => ty,
+            ABIOperand::Reg { ty, .. } | ABIOperand::Stack { ty, .. } => ty,
         }
     }
 }
 
-/// ABI-specific representation of the function result.
-#[derive(Copy, Clone, Debug)]
-pub(crate) enum ABIResult {
-    Void,
-    Reg {
-        /// Type of the result.
-        ty: WasmType,
-        /// Register to hold the result.
-        reg: Reg,
-    },
+/// Information about the [`ABIOperand`] information used in [`ABISig`].
+#[derive(Clone, Debug)]
+pub(crate) struct ABIOperands {
+    /// All the operands.
+    pub inner: SmallVec<[ABIOperand; 6]>,
+    /// All the registers used as operands.
+    pub regs: HashSet<Reg>,
+    /// Stack bytes used by the operands.
+    pub bytes: u32,
 }
 
-impl Default for ABIResult {
+impl Default for ABIOperands {
     fn default() -> Self {
-        Self::Void
+        Self {
+            inner: Default::default(),
+            regs: HashSet::with_capacity(0),
+            bytes: 0,
+        }
     }
 }
 
-impl ABIResult {
-    /// Create a register ABI result.
-    pub fn reg(ty: WasmType, reg: Reg) -> Self {
-        Self::Reg { ty, reg }
+/// ABI-specific representation of an [`ABISig`].
+#[derive(Clone, Debug, Default)]
+pub(crate) struct ABIResults {
+    /// The result operands.
+    operands: ABIOperands,
+}
+
+impl ABIResults {
+    /// Creates [`ABIResults`] from a slice of `WasmType`.
+    pub fn from<F>(returns: &[WasmType], mut map: F) -> Self
+    where
+        F: FnMut(&WasmType, u32) -> (ABIOperand, u32),
```

Comment:
Could we document what the `map` closure is supposed to do here? Was not immediately obvious to me when reading the code.

Reply:
Fixed!